### PR TITLE
Move native credentials lookup to the client library.

### PR DESCRIPTION
**Pull request by calavera**
This change extracts the native client logic from Docker
to use the one provided by the credentials library.

Tests remain here to verify that Docker's behavior with email
and whatnot still works, but the library also implements its own tests.

It would be awesome if someone could also review the changes in the library.
They should not affect the current behavior, but tests remain here to verify it.

https://github.com/docker/docker-credential-helpers/pull/15

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by calavera** [comment]
Oh, if you wonder why I want to do this, it's because I'm going to use that code in another CLI :grin:

File: hack/vendor.sh [hunk 1]
```diff
@@ -85,6 +85,7 @@ clone git github.com/ugorji/go 5abd4e96a45c386928ed2ca2a7ef63e2533e18ec
 clone git github.com/hashicorp/consul v0.5.2
 clone git github.com/boltdb/bolt v1.2.1
 clone git github.com/miekg/dns 75e6e86cc601825c5dbcd4e0c209eab180997cd7
+clone git github.com/docker/docker-credential-helpers 2ccf72627d3d89732442be0de61514c5f30701df https://github.com/calavera/docker-credential-helpers.git
```
Comment (thaJeztah) [inline hack/vendor.sh]: Should we create a tag for this?

Comment (calavera) [inline hack/vendor.sh]: first, I needs to land in master, it's in my branch waiting for a review :smile:

Comment (thaJeztah) [inline hack/vendor.sh]: oh! you're right, ignore ignore :)

**Comment by calavera** [comment]
Janky looks happy with the changes, but vendor will complain until docker/docker-credential-helpers#15 is merged. I do not have creds in that repo anymore :pensive:

**Comment by thaJeztah** [comment]
@calavera before we merge this, we should have a "maintainers" file added to that repository, in case changes are needed that we have enough people to make those go in smoothly

File: vendor/src/github.com/docker/docker-credential-helpers/client/client.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+func Get(program ProgramFunc, serverURL string) (*credentials.Credentials, error) {
+	cmd := program("get")
+	cmd.Input(strings.NewReader(serverURL))
+
+	out, err := cmd.Output()
+	if err != nil {
+		t := strings.TrimSpace(string(out))
+
+		// do not return an error if the credentials are not
+		// in the keyckain. Let the caller ask for new credentials.
```
Comment (runcom) [inline vendor/src/github.com/docker/docker-credential-helpers/client/client.go]: this comment is wrong now

**Comment by vdemeester** [comment]
@calavera you can update the vendor now :wink:

**Comment by calavera** [comment]
I finally figured out what was going on with the vendor script, it's a bug in `hack/.vendor-helpers`. It assumes that every package that starts with `github.com/docker/docker` belongs to the project and it was removing `github.com/docker/docker-credential-helpers` from the imported list. 

I included a patch in the vendoring commit:

https://github.com/docker/docker/pull/22988/files#diff-04aeacba9949f4aa0a8c6c812287ae5a

**Comment by thaJeztah** [comment]
Ah! Thanks for fixing that!

**Comment by thaJeztah** [comment]
LGTM

File: cliconfig/credentials/native_store.go [hunk 3]
```diff
@@ -127,70 +97,30 @@ func (c *nativeStore) storeCredentialsInStore(config types.AuthConfig) error {
...
+	creds, err := client.Get(c.programFunc, serverAddress)
 	if err != nil {
-		t := strings.TrimSpace(string(out))
-
-		// do not return an error if the credentials are not
-		// in the keyckain. Let docker ask for new credentials.
-		if t == errCredentialsNotFound.Error() {
+		if credentials.IsErrCredentialsNotFound(err) {
+			// do not return an error if the credentials are not
+			// in the keyckain. Let docker ask for new credentials.
```
Comment (runcom) [inline cliconfig/credentials/native_store.go]: I think the caller should decide here if not found should be ignored and err not found should be returned along with ret. But this is not blocking

**Comment by runcom** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐹

**Pull request merged**